ext/bcmath/libbcmath/src/recmul.c

```diff
@@ -277,34 +351,424 @@ static void bc_standard_mul(bc_num n1, size_t n1len, bc_num n2, size_t n2len, bc
+		return false;
+	}
+
+	while (calc_size > 0) {
+		calc_size /= BC_REC_MUL_DO_ADJUST_EXPO;
+		if (UNEXPECTED(calc_size == 1)) {
+			return true;
+		}
 	}
+	return false;
+}
 
+/*
+ * This is a function that is called recursively in divide-and-conquer calculations.
+ * if the size of n1 and n2 are both equal to calc_size, the calculation is performed
+ * without checking the number of digits.
+ * See bc_rec_mul's comment for the meaning of high, mid, and low.
+ */
+static void bc_rec_mul_recursive_fast(
+	const BC_UINT_T *n1, BC_UINT_T *n1_buf,
+	const BC_UINT_T *n2, BC_UINT_T *n2_buf,
+	size_t calc_size, BC_UINT_T *ret)
+{
 	/*
-	 * The last digit may carry over.
-	 * Also need to fill it to the end with zeros, so loop until the end of the string.
+	 * Size 2 is the smallest unit.
+	 * The calculation itself is skipped for the combination of size 2 and 0, and the array
+	 * size is adjusted to a multiple of 2, so the combination of 2 and 1 is impossible.
+	 * Therefore, the minimum unit calculation is always performed using this func.
 	 */
-	while (pend >= pptr) {
-		*pend-- = prod_uint[i] % BASE;
-		prod_uint[i] /= BASE;
+	if (calc_size == 2) {
+		BC_UINT_T low = n1[0] * n2[0];
+		BC_UINT_T high = n1[1] * n2[1];
+		ret[0] = low;
+		ret[1] = low + high - (n1[1] - n1[0]) * (n2[1] - n2[0]);
+		ret[2] = high;
+		ret[3] = 0;
+		return;
+	}
+
+	size_t i;
+	size_t half_size = calc_size / 2;
+
+	/* low */
+	BC_UINT_T *low = ret;
+	bc_rec_mul_recursive_fast(n1, n1_buf, n2, n2_buf, half_size, low);
+	/* high */
+	BC_UINT_T *high = ret + calc_size;
+	bc_rec_mul_recursive_fast(n1 + half_size, n1_buf, n2 + half_size, n2_buf, half_size, high);
+
+	/* prepare mid */
+	for (i = 0; i < half_size; i++) {
+		size_t high_i = i + half_size;
+		n1_buf[i] = n1[high_i] - n1[i];
+		n2_buf[i] = n2[high_i] - n2[i];
+	}
+
+	/* mid */
+	BC_UINT_T *mid = high + calc_size;
+	bc_rec_mul_recursive_fast(n1_buf, n1_buf + half_size, n2_buf, n2_buf + half_size, half_size, mid);
+
+	/*
+	 * Add to ret.
+	 * From 0 to half_size, adding it to ret will affect the low value, so add it later.
+	 */
+	for (i = 0; i < half_size; i++) {
+		mid[i] -= low[i] + high[i];
+	}
+	for (; i < calc_size; i++) {
+		ret[i + half_size] += low[i] + high[i] - mid[i];
+	}
+	for (i = 0; i < half_size; i++) {
+		ret[i + half_size] -= mid[i];
+	}
+
+	/*
+	 * Perform digit adjustment to avoid overflow.
+	 */
+	if (UNEXPECTED(bc_rec_mul_near_overflow(calc_size))) {
+		for (i = 0; i < calc_size * 2; i++) {
+			bc_digits_adjustment_single(ret + i);
+		}
+	}
+}
+
+/*
+ * This is a function that is called recursively in divide-and-conquer calculations.
+ * Compute the three multiplications of Karatsuba-algorithm as high, mid, and low, respectively.
+ * e.g. 1234 * 5678
+ * low = 34 * 78
+ * high = 12 * 56
+ * mid = (12 - 34) * (56 - 78)
+ * ret = high * 100^2 + (high + low - mid) * 100 + low
+ * This function differs from "fast" in that it does things like check the number of digits.
```

Comment:
I always find these things scary, especially if they're complex...

Reply:
That meens about the risks of skipping checks in the fast version?

Reply:
Yeah, the comment is about the trade-off between more complex code (i.e. without checks) vs simpler code (i.e. always with checks). If it's a notable speedup to avoid the checks then we should keep it but we still need very thorough testing for all of this to justify the complexity.

Reply:
I changed to using ZEND_ASSERT, although it is slightly slower.

[5b75112](https://github.com/php/php-src/pull/14376/commits/5b75112d53d584da4d81b59a88f7761b7ebb4c24)